pod and comments: Note escape vs quote

khwilliamson opened the pull request:
Fixes #15221

The documentation and comments was misleading about conflating quoting a metacharacter and escaping it.  Since \Q stands for quote, we have to  continue to use that terminology.  This commit clarifies that the two terms are often equivalent.

* This set of changes does not require a perldelta entry.

khwilliamson committed:
perlintro: Define metacharacter before using the term

This adds a bit of text about metacharacters that was missing from this
introductory pod.

guest20 commented on pod/perlfunc.pod:
```diff
@@ -6536,6 +6536,11 @@ the C<\Q> escape in double-quoted strings.
 
+The motivation behind this is to make all characters in EXPR have their
+literal meanings.  Otherwise any metacharacters in it could trigger
+their "magic" actions.  The characters are said to be "quoted" or
```
I don't think metachacters have "actions", maybe they "behave", but the definitely don't "act"

khwilliamson replied:
changed as suggested

Grinnz commented on pod/perlintro.pod:
```diff
@@ -584,10 +584,33 @@ the meantime, here's a quick cheat sheet:
-want to match on, where "thing" means either a literal character, one
-of the metacharacters listed above, or a group of characters or
-metacharacters in parentheses.
+Note that in the above, C<$> doesn't mean a dollar sign.  Similarly
```
Perhaps slightly clearer to say, "`$` doesn't match a dollar sign."

Grinnz replied:
And if so, "mean" could be replaced with "match" to clarify some other statements in this section.

khwilliamson replied:
Good points;  check out the changed version

Grinnz commented on pod/perlre.pod:
```diff
@@ -1348,17 +1348,17 @@ their punctuation character equivalents, however at the trade-off that you
-once used in a common idiom to disable or quote the special meanings
-of regular expression metacharacters in a string that you want to
-use for a pattern. Simply quote all non-"word" characters:
+=head2 Quoting (or escaping) metacharacters
```
Changing headings is perilous, but I found no instances of this being linked to directly on CPAN.

khwilliamson replied:
I removed the change

khwilliamson committed:
pod and comments: Note escape vs quote

Fixes #15221

The documentation and comments were misleading about conflating quoting
a metacharacter and escaping it.  Since \Q stands for quote, we have to
continue to use that terminology.  This commit clarifies that the two
terms are often equivalent.

This also adds detail about quotemeta and \Q.

Grinnz commented on pod/perlre.pod:
```diff
@@ -1350,25 +1350,36 @@ X</p> X<p modifier>
-Today it is more common to use the C<L<quotemeta()|perlfunc/quotemeta>>
+(If C<use locale> is in effect, then this depends on the current locale.)
+
+This template used to be a common paradigm, but these days it is more
```
It's not really the purpose of this PR, but I think it's a little misleading to say that quotemeta "disables metacharacters special meanings" without also saying that it actually does a very similar thing to the example just above, without being locale dependent for ASCII characters. i.e. it's exactly equivalent to `s/([^A-Za-z_0-9])/\\$1/g` or `s/(\W)/\\$1/ag` for ASCII characters (with behavior on non-ASCII characters dependent on the unicode_strings feature).

Grinnz replied:
With the additional very useful feature that it can be interpolated with `\Q` and `\E` rather than applied to the whole pattern.

khwilliamson replied:
@Grinnz please check if new wording is ok

Grinnz reviewed (approved):
Otherwise I think this is an excellent improvement

jkeenan reviewed (commented):
I'm generally in favor of the changes presented in this p.r.  But there is a lot in it, and I'd particularly like extra eyeballs on the [changes in pod/perlre.pod](https://github.com/Perl/perl5/pull/23264/files#diff-1910b59f34148cf03dfffeaaad474b00434140560a2cbd8cf5449ea84d364f8e).

Grinnz commented on pod/perlre.pod:
```diff
@@ -1350,24 +1350,60 @@ X</p> X<p modifier>
+
+    quotemeta $pattern;
+
+does either of these:
```
I'm not quite sure this is correct, it's not quite how the quotemeta docs describe it, but it sounds complex. In particular it says that ASCII characters are always subject to the ASCII style escaping regardless of unicode_strings or the utf8 flag. Maybe better to say it does similar things, give the basics, and link to the quotemeta docs for details.

khwilliamson replied:
You're right, I thought I knew what was going on, but didn't.  I have since RTFM, and the code, and I'm much happier with the newest iteration.

I also added a commit which changes a heading as I originally did (and which I think reads better) and has a workaround in case someone did link to the old wording

Grinnz replied:
LGTM

khwilliamson committed:
perlre: Simplify some text

By adding a word to a =heading, a sentence can be removed, and is
clearer.

But since some pod somewhere  may link to that heading, a section at the
end is added with the old name, and pointing to the new one.

Dan Book searched and found no instances in CPAN of the old heading
being linked to, but this guarantees that nothing breaks.
